File: ciel/evolution_hydra/FitnessEvaluator.ts
/**
 * ═══════════════════════════════════════════════════════════════
 * FITNESS EVALUATOR — Calcule la fitness réelle d'un génome
 * ═══════════════════════════════════════════════════════════════
 *
 * La fitness n'est plus un chiffre arbitraire. Elle combine :
 * - Performance de l'agent dans sa dernière mission (40%)
 * - Satisfaction utilisateur estimée (30%)
 * - Efficacité des ressources (coût, latence) (20%)
 * - Pression évolutive (karma, ombre) (10%)
 *
 * Branché à ImperialCycle dans Phase 2.
 */

import type { UnifiedGenome } from './UnifiedGenome.js';

export interface HydraContext {
  /** Score de succès de la dernière mission (0-1). */
  taskSuccess: number;
  /** Satisfaction utilisateur dérivée (0-1). */
  userSatisfaction: number;
  /** Coût relatif (0 = gratuit, 1 = très cher). */
  costRatio: number;
  /** Latence relative (0 = instant, 1 = très lent). */
  latencyRatio: number;
  /** Pression évolutive (karma négatif ou stagnation). */
  evolutionaryPressure: number;
  /** Métadonnées additionnelles (optionnel). */
  metadata?: Record<string, unknown>;
}

export interface IFitnessEvaluator {
  evaluate(genome: UnifiedGenome, context?: HydraContext): number;
}

export const DEFAULT_FITNESS_WEIGHTS = {
  taskSuccess: 0.40,
  userSatisfaction: 0.30,
  costEfficiency: 0.20,
  evolutionaryPressure: 0.10,
} as const;

/**
 * Implémentation par défaut : combine les dimensions du contexte
 * avec un bonus pour les génomes équilibrés (entropie de Shannon des gènes).
 */
export class DefaultFitnessEvaluator implements IFitnessEvaluator {
  private weights = { ...DEFAULT_FITNESS_WEIGHTS };

  public evaluate(genome: UnifiedGenome, context?: HydraContext): number {
    // Si pas de contexte, retomber sur l'historique interne
    if (!context) {
      return this.fromHistory(genome);
    }

    const taskScore = Math.max(0, Math.min(1, context.taskSuccess));
    const userScore = Math.max(0, Math.min(1, context.userSatisfaction));
    const costScore = Math.max(0, Math.min(1, 1 - context.costRatio));  // bas coût = bon
    const latencyScore = Math.max(0, Math.min(1, 1 - context.latencyRatio));
    const efficiencyScore = (costScore + latencyScore) / 2;
    const pressureScore = Math.max(0, Math.min(1, context.evolutionaryPressure));

    const fitness =
      taskScore * this.weights.taskSuccess +
      userScore * this.weights.userSatisfaction +
      efficiencyScore * this.weights.costEfficiency +
      pressureScore * this.weights.evolutionaryPressure;

    return Math.max(0, Math.min(1, fitness));
  }

  /**
   * Fallback : fitness dérivée de l'historique et de la diversité génétique.
   */
  private fromHistory(genome: UnifiedGenome): number {
    if (genome.fitnessHistory.length === 0) return genome.fitness || 0.5;
    const recent = genome.fitnessHistory.slice(-10);
    const avg = recent.reduce((a, b) => a + b, 0) / recent.length;
    // Bonus de diversité : entropie des gènes du chromosome principal
    const diversityBonus = this.computeDiversityBonus(genome) * 0.05;
    return Math.max(0, Math.min(1, avg + diversityBonus));
  }

  /**
   * Entropie de Shannon normalisée des valeurs de gènes (0-1).
   * Une diversité élevée indique un génome "exploratoire".
   */
  private computeDiversityBonus(genome: UnifiedGenome): number {
    const sample = genome.g_behavior.slice(0, 50);
    if (sample.length === 0) return 0;
    const buckets = new Array(10).fill(0);
    for (const gene of sample) {
      const idx = Math.min(9, Math.floor(gene.value * 10));
      buckets[idx]++;
    }
    const total = sample.length;
    let entropy = 0;
    for (const count of buckets) {
      if (count > 0) {
        const p = count / total;
        entropy -= p * Math.log2(p);
      }
    }
    // Max entropy pour 10 buckets = log2(10) ≈ 3.32
    return entropy / Math.log2(10);
  }

  public getWeights(): typeof this.weights {
    return { ...this.weights };
  }
}

File: ciel/evolution/fitness_evaluator.py
from __future__ import annotations

import math
from dataclasses import dataclass, field
from typing import Any


@dataclass(slots=True)
class HydraContext:
    task_success: float = 0.0
    user_satisfaction: float = 0.0
    cost_ratio: float = 0.0
    latency_ratio: float = 0.0
    evolutionary_pressure: float = 0.0
    metadata: dict[str, Any] | None = None


FITNESS_WEIGHTS = {
    "task_success": 0.40,
    "user_satisfaction": 0.30,
    "cost_efficiency": 0.20,
    "evolutionary_pressure": 0.10,
}


class DefaultFitnessEvaluator:
    def __init__(self) -> None:
        self._weights = dict(FITNESS_WEIGHTS)

    def evaluate(self, genome: Any, context: HydraContext | None = None) -> float:
        if context is None:
            return self._from_history(genome)
        task_score = max(0.0, min(1.0, context.task_success))
        user_score = max(0.0, min(1.0, context.user_satisfaction))
        cost_score = max(0.0, min(1.0, 1.0 - context.cost_ratio))
        latency_score = max(0.0, min(1.0, 1.0 - context.latency_ratio))
        efficiency_score = (cost_score + latency_score) / 2.0
        pressure_score = max(0.0, min(1.0, context.evolutionary_pressure))
        fitness = (
            task_score * self._weights["task_success"]
            + user_score * self._weights["user_satisfaction"]
            + efficiency_score * self._weights["cost_efficiency"]
            + pressure_score * self._weights["evolutionary_pressure"]
        )
        return max(0.0, min(1.0, fitness))

    def _from_history(self, genome: Any) -> float:
        history = getattr(genome, "fitness_history", [])
        fitness = getattr(genome, "fitness", 0.5)
        if not history:
            return fitness
        recent = history[-10:]
        avg = sum(recent) / len(recent)
        bonus = self._compute_diversity_bonus(genome) * 0.05
        return max(0.0, min(1.0, avg + bonus))

    def _compute_diversity_bonus(self, genome: Any) -> float:
        behavior = getattr(genome, "g_behavior", [])
        sample = behavior[:50]
        if not sample:
            return 0.0
        buckets = [0] * 10
        for gene in sample:
            val = getattr(gene, "value", 0.5)
            idx = min(9, int(val * 10))
            buckets[idx] += 1
        total = len(sample)
        entropy = 0.0
        for count in buckets:
            if count > 0:
                p = count / total
                entropy -= p * math.log2(p)
        return entropy / math.log2(10)

    def get_weights(self) -> dict[str, float]:
        return dict(self._weights)

    def process(self, input_data: Any) -> dict[str, Any]:
        if isinstance(input_data, dict):
            ctx = HydraContext(
                task_success=input_data.get("task_success", 0.0),
                user_satisfaction=input_data.get("user_satisfaction", 0.0),
                cost_ratio=input_data.get("cost_ratio", 0.0),
                latency_ratio=input_data.get("latency_ratio", 0.0),
                evolutionary_pressure=input_data.get("evolutionary_pressure", 0.0),
            )
            return {
                "weights": self._weights,
                "evaluator_type": "DefaultFitnessEvaluator",
            }
        return {"error": "No context provided"}
